src/nspawn/nspawn-oci.c

```diff
@@ -0,0 +1,2350 @@
+                }
+
+                m = (uint64_t) json_variant_unsigned(k);
+
+                if ((uintmax_t) m != json_variant_unsigned(k)) {
+                        json_log(v, flags, 0, "pids limit out of range, refusing.");
+                        return -EINVAL;
+                }
+        }
+
+        r = settings_allocate_properties(s);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_append(s->properties, "(sv)", "TasksMax", "t", m);
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        return 0;
+}
+
+static int oci_resources(const char *name, JsonVariant *v, JsonDispatchFlags flags, void *userdata) {
+
+        static const JsonDispatch table[] = {
+                { "devices",        JSON_VARIANT_ARRAY,  oci_cgroup_devices,  0, 0 },
+                { "memory",         JSON_VARIANT_OBJECT, oci_cgroup_memory,   0, 0 },
+                { "cpu",            JSON_VARIANT_OBJECT, oci_cgroup_cpu,      0, 0 },
+                { "blockIO",        JSON_VARIANT_OBJECT, oci_cgroup_block_io, 0, 0 },
+                { "hugepageLimits", JSON_VARIANT_ARRAY,  oci_unsupported,     0, 0 },
+                { "network",        JSON_VARIANT_OBJECT, oci_unsupported,     0, 0 },
+                { "pids",           JSON_VARIANT_OBJECT, oci_cgroup_pids,     0, 0 },
+                { "rdma",           JSON_VARIANT_OBJECT, oci_unsupported,     0, 0 },
+                {}
+        };
+
+        return json_dispatch(v, table, oci_unexpected, flags, userdata);
+}
+
+static bool sysctl_key_valid(const char *s) {
+        bool dot = true;
+
+        /* Note that we are a bit stricter here than in systemd-sysctl, as that inherited semantics from the old sysctl
+         * tool, which were really weird (as it swaps / and . in both ways) */
+
+        if (isempty(s))
+                return false;
+
+        for (; *s; s++) {
+
+                if (*s <= ' ' || *s >= 127)
+                        return false;
+                if (*s == '/')
+                        return false;
+                if (*s == '.') {
+
+                        if (dot) /* Don't allow two dots next to each other (or at the beginning) */
+                                return false;
+
+                        dot = true;
+                } else
+                        dot = false;
+        }
+
+        if (dot) /* don't allow a dot at the end */
+                return false;
+
+        return true;
+}
+
+static int oci_sysctl(const char *name, JsonVariant *v, JsonDispatchFlags flags, void *userdata) {
+        Settings *s = userdata;
+        JsonVariant *k, *w;
+        int r;
+
+        assert(s);
+
+        JSON_VARIANT_OBJECT_FOREACH(k, w, v) {
+                const char *n, *m;
+
+                if (!json_variant_is_string(w)) {
+                        json_log(v, flags, 0, "sysctl parameter is not a string, refusing.");
+                        return -EINVAL;
+                }
+
+                assert
```

Comment:
So, the json parser already filtered out these errors here, as we required that the sysctl object is actually an object. And an object has to have every second element as a string (because that's the fields key name). And every other second element can be something else, but we checked that it is a string a few lines above. Hence, when pulling out the two strings from `k` and `v` here we know for sure in both cases that it is a string. Or to say this differently: if the variant here is not an object, then json_dispatch() already has complained for us via json_log(). If the `k` is not a string then the parser should have complained even earlier already, also with json_log(), and if `v` is not a string we log with json_log() right here, so all grounds should be covered: we log in all bad cases, always friendly, and the assert_se() is only used to catch invalid expectations, i.e. if we cannot get a string for something that already was veryified to be a string.

Reply:
oh, and the oci spec says to refuse unknown fields, and says that the oci version at the top of the json data should be updated on extensions